Add support for Text fragment feature (#1545)

thiru-appitap opened the pull request:
Text Fragment feature implementation pull request. The feature follows the published URL Fragment Text Directives specification (https://wicg.github.io/scroll-to-text-fragment/). 

(base) lychee % lychee -vv --include-text-fragments  https://developer.mozilla.org/en-US/docs/Web/URI/Fragment/Text_fragments
...
[DEBUG] tdirective: "text=From%20the%20foregoing%20remarks%20we%20may%20gather%20an%20idea%20of%20the%20importance
[DEBUG] status: Completed
[DEBUG] result: "From the foregoing remarks we may gather an idea of the importance"
[200] https://mdn.github.io/css-examples/target-text/index.html#:~:text=From%20the%20foregoing%20remarks%20we%20may%20gather%20an%20idea%20of%20the%20importance
[DEBUG] tdirective: "text=linked%20URL,-'s%20format"
[DEBUG] status: Completed
[DEBUG] result: "linked URL"
[DEBUG] tdirective: "text=Deprecated-,attributes,attribute"
[DEBUG] status: Completed
[DEBUG] result: "attributes     charset Deprecated   Hinted at the character encoding of the linked URL.   Note:This attribute"
[200] https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=linked%20URL,-'s%20format&text=Deprecated-,attributes,attribute
[DEBUG] tdirective: "text=downgrade:-,The%20Referer,be%20sent,-to%20origins"
[DEBUG] status: Completed
[DEBUG] result: "The Referer header will not be sent"
[200] https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=downgrade:-,The%20Referer,be%20sent,-to%20origins
[DEBUG] tdirective: "text=linked%20URL,defining%20a%20value"
[DEBUG] status: Completed
[DEBUG] result: "linked URL as a download. Can be used with or without a filename value:    Without a value, the browser will suggest a filename/extension, generated from various sources:   The Content-Disposition HTTP header  The final segment in the URL path  The media type (from the Content-Type header, the start of a data: URL, or Blob.type for a blob: URL)     filename: defining a value"
[200] https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=linked%20URL,defining%20a%20value
...     

If the fragment directive is not found, a TextDirectiveNotFound error will be returned.

Below changes are completed:
1. Fragment Directive parser uses fancy-regex
- this package was added as a dependency
2. a new flag, include-text-fragments is added to support the feature
- this is a deviation from the original feature request (which asked for using the text-fragments flag itself) 
3. Fragment (Text) Directive feature is tested on LTR sites only
4. new UrlExt trait is implemented to enhance Url's to support Fragment Directive
5. Support for multiple text fragment directives (for example, #:~:text=linked%20URL,-'s%20format&text=Deprecated-,attributes,attribute)
5. tests are added for validating the feature
6. cargo clippy & cargo tests were executed

thiru-appitap commented:
I missed to run the clippy across the test modules - the related lint failure issues are now fixed and ready for review!

mre commented on lychee-lib/src/checker/website.rs:
```diff
@@ -86,9 +93,37 @@ impl WebsiteChecker {
         match self.reqwest_client.execute(request).await {
-            Ok(ref response) => Status::new(response, self.accepted.clone()),
+            Ok(response) => {
+                let mut status = Status::new(&response, self.accepted.clone());
+                if self.validate_text_fragments && has_fragment_directive {
+                    if let Ok(res) = response.text().await {
+                        info!("checking fragment directive...");
+                        if let Some(fd) = req_url.fragment_directive() {
+                            info!("directive: {:?}", fd.text_directives);
+                            match fd.check(&res) {
+                                Ok(stat) => {
+                                    status = stat;
+                                }
+                                Err(e) => {
+                                    return e.into();
+                                }
+                            }
+                        }
+                    }
+                }
+                status
```
Maybe we can move that into a function/method?

thiru-appitap replied:
the code is modified to move this into a separate function - as well, bulk of the logic is abstracted at the textfrag crate itself so that the websitechecker will deal with only error responses from the text fragment checker function.

mre commented on lychee-lib/src/client.rs:
```diff
@@ -841,6 +846,65 @@ mod tests {
+        assert!(res.status().is_success());
+
+        // start with suffix
+        println!("\ntesting start with suffix...");
```
you'll probably remove the `println!`s right?

thiru-appitap replied:
yes - the println's are removed now

mre commented on lychee-lib/src/extract/html/fragdirtok.rs:
```diff
@@ -0,0 +1,777 @@
+
+use crate::types::TextDirective;
+
+const BLOCK_ELEMENTS: &[&str] = &[
```
This is a long list. Does that mean we'd have to maintain the HTML keywords here? Maybe we can avoid that as it would be an uphill battle.

thiru-appitap replied:
I am yet to explore alternative approach to this and am open for suggestion - for now, this list is retained (reserving this change for a near future commit)

mre commented on lychee-lib/src/extract/html/fragdirtok.rs:
I like that this functionality is isolated in its own module. But it's a looot of code. 😅 Not sure what to do here, but at least the ratio of code/tests could be improved.

mre replied:
Perhaps we could move it into a separate crate or use an upstream crate for that? I think it would be a nice library to maintain individually as more applications could profit from it

thiru-appitap replied:
text fragment is now moved into its own crate, textfrag, in the lychee main tree itself - as I am new to the ecosystem, I'll need help in moving this into a separate crate - please suggest!

mre commented on lychee-lib/src/extract/html/fragdirtok.rs:
```diff
@@ -0,0 +1,777 @@
+        let mut all_directives_found = false;
+        let directive = td.directive.borrow();
+
+        'directive_loop: while !all_directives_found {
```
The labels make it quite hard to read. Have you considered any alternatives?

thiru-appitap replied:
yes - the code logic was cumbersome and I've rewritten this now - please share your comments!

mre commented on lychee-lib/src/types/text_fragment.rs:
```diff
@@ -0,0 +1,562 @@
+pub(crate) const FRAGMENT_DIRECTIVE_DELIMITER: &str = ":~:";
+pub(crate) const TEXT_DIRECTIVE_DELIMITER: &str = "text=";
+
+pub(crate) const TEXT_DIRECTIVE_REGEX: &str = r"(?s)^text=(?:\s*(?P<prefix>[^,&-]*)-\s*[,$]?\s*)?(?:\s*(?P<start>[^-&,]*)\s*)(?:\s*,\s*(?P<end>[^,&-]*)\s*)?(?:\s*,\s*-(?P<suffix>[^,&-]*)\s*)?$";
```
Where does that regex come from? Did you write it yourself? If there's an "official" regex for those text fragments, we could perhaps add a link to the reference.

thiru-appitap replied:
this regex was built by myself - I started searching for any equivalent but couldn't land on one when I started this implementation - I am open to revisit this.

In fact, i want to replace regex with a simple parser for this requirement - the specification is not particular about the order of the directives, whereas the regex assumes the order imperatively - with a parser, we might be able to get away from the ordering constraints.

mre commented on lychee-lib/src/utils/url.rs:
```diff
@@ -23,10 +26,61 @@ pub(crate) fn find_links(input: &str) -> impl Iterator<Item = linkify::Link> {
 }
 
+/// Fragment Directive feature trait
+/// we will use the extension trait pattern to extend the Url to support Text Fragment feature
```
Nice idea!

thiru-appitap replied:
Thank you - this has now moved into textfrag crate

mre commented on lychee-lib/src/utils/url.rs:
```diff
@@ -23,10 +26,61 @@ pub(crate) fn find_links(input: &str) -> impl Iterator<Item = linkify::Link> {
+}
+
+impl UrlExt for Url {
+    /// Returns whether the URL has fragment directive or not
```
```suggestion
    /// Checks whether the URL has fragment directive or not
```

thiru-appitap replied:
Have addressed this is in the textfrag::utils::url[:16]

mre reviewed (changes requested):
Left some comments. I like the overall structure. Good work so far!

mre commented:
@Thiru, any updates? Let me know in case you need any help. 😃

thiru-appitap commented:
> @thiru, any updates? Let me know in case you need any help. 😃

@mre, I will submit the fixes by this weekend for your review. Earlier I started moving the feature into a separate crate, while addressing your review comments, but got pulled into few other tasks and so couldn't get back earlier :-(.

mre commented:
Thanks, sounds good!

thiru-appitap commented:
> Thanks, sounds good!

@mre my apologies for the delay - while refactoring, into a separate crate, I encountered few corner case issues and it took more time than planned - running the tests now and hoping to re-raise the pull-request by tomorrow!

thiru-appitap commented on lychee-lib/src/utils/url.rs:
```diff
@@ -23,10 +26,61 @@ pub(crate) fn find_links(input: &str) -> impl Iterator<Item = linkify::Link> {
+/// Fragment Directive feature trait
+/// we will use the extension trait pattern to extend the Url to support Text Fragment feature
+pub(crate) trait UrlExt {
+    /// Checks if the url has a fragment and if the fragment is has the fragment directive delimiter embedded
```
Updated the comments section (with slight rewording) in the textfrag crate

thiru-appitap commented on lychee-lib/src/utils/url.rs:
```diff
@@ -23,10 +26,61 @@ pub(crate) fn find_links(input: &str) -> impl Iterator<Item = linkify::Link> {
 }
 
+/// Fragment Directive feature trait
+/// we will use the extension trait pattern to extend the Url to support Text Fragment feature
```
Incorporated this comment in to the textfrag::utils::url file

thiru-appitap commented on lychee-lib/src/utils/url.rs:
```diff
@@ -23,10 +26,61 @@ pub(crate) fn find_links(input: &str) -> impl Iterator<Item = linkify::Link> {
+        );
+        match url {
+            Ok(url) => {
+                eprintln!(
```
Modified the tests to have asserts (instead of the manual checks that was earlier)

thiru-appitap commented:
I am committing the changes into my fork and will initiate a pull-request shortly!

thiru-appitap commented:
@mre request your help in addressing the CI / publish-check failure - please recommend if I've to make any changes on my end - thank you!

mre commented on lychee-lib/src/checker/website.rs:
```diff
@@ -85,10 +94,51 @@ impl WebsiteChecker {
 
+    fn check_text_fragments(site_data: &str, url: &Url, mut status: Status) -> Status {
+        let res = check_text_fragments(site_data, url);
+        if res.is_err() {
```
You don't need the if condition here. You're matching on `res.err()` below, and in the case where res.err() is not set, it will be `None`. This case is already covered in your `_res` match arm since `_res` is just a placeholder, which also includes the value being `None`.

thiru-appitap replied:
Yes - I've updated the code

mre commented on textfrag/src/extract/mod.rs:
```diff
@@ -0,0 +1,5 @@
+// mod blockcontent;
```
I think this line can be cleaned up.

thiru-appitap replied:
clean up done - will be more careful going forward

mre commented on textfrag/src/extract/mod.rs:
```diff
@@ -0,0 +1,5 @@
+// mod blockcontent;
+mod fragdirtok;
+
+// pub use blockcontent::BlockElementContent;
```
Can be cleaned up.

thiru-appitap replied:
done

mre commented on textfrag/src/lib.rs:
```diff
@@ -0,0 +1,21 @@
+    site_data: &str,
+    url: &Url,
+) -> Result<FragmentDirectiveStatus, FragmentDirectiveError> {
+    log::info!("checking fragment directive...");
```
Not sure if we need this info message? Or if so, maybe `log::debug` or `log::trace`.

thiru-appitap replied:
I agree - this is not necessary any more - cleaning it up!

mre commented on textfrag/src/lib.rs:
```diff
@@ -0,0 +1,21 @@
+) -> Result<FragmentDirectiveStatus, FragmentDirectiveError> {
+    log::info!("checking fragment directive...");
+    if let Some(fd) = url.fragment_directive() {
+        log::info!("directives: {:?}", fd.text_directives());
```
could be `log::debug`.

thiru-appitap replied:
changed to log::debug

mre commented on textfrag/src/types/blockcontent.rs:
```diff
@@ -0,0 +1,216 @@
+/// HTML Block element content
```
I think the documentation could be a bit more extensive.

E.g.

```rust
//! HTML Block Element Content
//!
//! Finds a search string in a block of content.
//! Returns a pair of `start` and `end` offsets on the block content
//!
//! # Examples
//!
//! ```rust
//! todo!("example here");
//! ```
```

mre replied:
Examples get executed as unit tests (so-called doctests).
I would also love to see a bunch of normal unit tests at the end of the file. 😉

thiru-appitap replied:
updating the file to include more documentation - will also add unit tests

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+/// Fragment Directive object is collection of text fragments in the URL's fragment
+/// :~: is the fragment directive delmiter, followed by list of `[TextDirective]`
```
You can use `//!` to make this comment the official module description:

```suggestion
//! Fragment Directive object is collection of text fragments in the URL's fragment
//! :~: is the fragment directive delmiter, followed by list of `[TextDirective]`
```

thiru-appitap replied:
changes done!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+        self.text_directives.borrow_mut().to_owned()
+    }
+
+    // Extract Text Directives, from the Url fragment string
```
Use three slashes for function/method comments. This way they show up in the docs later. 😎 

```suggestion
    /// Extract Text Directives, from the Url fragment string
```

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+}
+
+impl FragmentDirective {
+    pub fn text_directives(&self) -> Vec<TextDirective> {
```
Ideally, I'd like to have a comment for every method. That's because a human description can add additional context that a function header might not miss.

For example, in this case:

```suggestion
    /// Returns all the defined text directives
    ///
    /// Note that this is a `Vector` and not a `HashSet`, because the directives get returned
    /// in the order they were defined. The order matters because...
    pub fn text_directives(&self) -> Vec<TextDirective> {
```

This is just an example. I'm actually not sure if the order matters. If it doesn't then a `HashSet` would be the more precise type.

thiru-appitap replied:
order will not matter as each of the object - i'm checking the recommendation!

thiru-appitap replied:
on second thoughts, even though the order of the text directives does not matter, I don't think we'll have any major benefits in picking up HashSet over the vectors - below are my reasonings:
1. the number of text directives will be always limited - it shall not be more than a handful directives
2. tokenizer processes the directives in an iterative manner and since there is no lookup requirements in here the HashSet advantage of O(1) operation is not that significant

given these thoughts, I am inclined to retain the vector over the HashSet.

please do let me know if I am wrong in the above understanding!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+                if let Ok(text_directive) = text_directive {
+                    text_directives.push(text_directive);
+                } else {
+                    // WARN: <log> & continue with other directives!
```
```suggestion
```

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+            return Ok(text_directives);
+        }
+
+        // WARN: <log>
```
```suggestion
```

thiru-appitap replied:
cleaned now!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+    /// Return an error if
+    /// - No match is found
+    /// - Suffix error (spec instructs the fragment SHALL be upto **Suffix** and this error is returned if this condition is violated)
+    ///
+    ///
```
```suggestion
```

thiru-appitap replied:
removed he extra comments line!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+    ///
+    /// # Errors
+    /// - `TextDirectiveNotFound`, if text directive match fails
+    // fn check_fragment_directive(&self, buf: &str) -> Result<TextFragmentStatus, TextFragmentError> {
```
```suggestion
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+        buf: &str,
+    ) -> Result<FragmentDirectiveStatus, FragmentDirectiveError> {
+        let mut map = HashMap::new();
+        let fd_checker = FragmentDirectiveTokenizer::new(self.text_directives()); // self.text_directives().clone());
```
```suggestion
        let fd_checker = FragmentDirectiveTokenizer::new(self.text_directives());
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+        for td in &tok.sink.get_text_directives() {
+            let directive = td.raw_directive().to_string();
+            log::debug!("text directive: {:?}", directive);
+            println!("text directive: {:?}", directive);
```
```suggestion
```

thiru-appitap replied:
removed unnecessary println!'s

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+
+            let _status = status.to_string();
+            log::debug!("search status: {:?}", status);
+            println!("search status: {:?}", status);
```
```suggestion
```

thiru-appitap replied:
removed unnecessary println!'s

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+
+            let res_str = td.get_result_str();
+            log::debug!("search result: {:?}", res_str);
+            println!("search result: {:?}", res_str);
```
```suggestion
```

thiru-appitap replied:
removed unnecessary println!'s

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+
+            let results = fd.check(MULTILINE_INPUT);
+            assert!(results.is_ok());
+            // assert_eq!(results.len(), 1);
+            // assert_eq!(results[FRAGMENT], Ok(TextDirectiveStatus::Completed));
```
```suggestion
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+
+            let res = fd.check(MULTILINE_INPUT);
+            assert!(res.is_ok());
+            // assert_eq!(res.len(), 1);
+            // assert_eq!(res[FRAGMENT], TextDirectiveStatus::Completed);
```
```suggestion
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+
+            let results = fd.check(MULTILINE_INPUT);
+            assert!(results.is_ok());
+            // assert_eq!(results.len(), 1);
+            // assert_eq!(results[FRAGMENT], TextDirectiveStatus::Completed);
```
```suggestion
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+
+            let results = fd.check(MULTILINE_INPUT);
+            assert!(results.is_ok());
+            // assert_eq!(results.len(), 1);
+            // assert_eq!(results[FRAGMENT], TextDirectiveStatus::Completed);
```
```suggestion
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+
+            let results = fd.check(MULTILINE_INPUT);
+            assert!(results.is_ok());
+            // assert_eq!(results.len(), 1);
+            // assert_eq!(results[FRAGMENT], TextDirectiveStatus::Completed);
```
```suggestion
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+
+            let results = fd.check(MULTILINE_INPUT);
+            assert!(results.is_ok());
+            // assert!(results.len() == 1);
+            // assert_eq!(results[FRAGMENT], TextDirectiveStatus::Completed);
```
```suggestion
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/directive/textdirective.rs:
```diff
@@ -0,0 +1,448 @@
+
+    /// [Internal] the use of regular expression does not comply with the specification
+    /// To be used for testing purposes only
+    fn _check(&self, input: &str) -> Result<FragmentDirectiveStatus, TextFragmentError> {
```
Using underscores for internal methods is rather uncommon. It's clear enought that there's no `pub` in front.

```suggestion
    fn check(&self, input: &str) -> Result<FragmentDirectiveStatus, TextFragmentError> {
```

thiru-appitap replied:
changed the function name and also to suppress the warning, i've included the #[allow(dead_code)] directive

thiru-appitap replied:
renamed the function per feedback

mre commented on textfrag/src/types/mod.rs:
```diff
@@ -0,0 +1,13 @@
+mod error;
+mod status;
+mod url;
+// mod frag_directive;
```
```suggestion
```

thiru-appitap replied:
clean-up done!

mre commented on textfrag/src/types/mod.rs:
```diff
@@ -0,0 +1,13 @@
+pub use error::*;
+pub use status::*;
+pub use url::*;
+// pub use frag_directive::*;
```
```suggestion
```

thiru-appitap replied:
clean-up done

mre commented on textfrag/src/types/status.rs:
```diff
@@ -0,0 +1,78 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            FragmentDirectiveStatus::Ok => write!(f, "Ok"),
+            // FragmentDirectiveStatus::PartialOk(m) => write!(f, "Partial Ok {:?}", m),
+            // FragmentDirectiveStatus::Error(e) => write!(f, "Error: {:?}", e),
```
What about these?

thiru-appitap replied:
take care of now - I've made sure that all the unwanted code is now cleaned entirely

mre commented on textfrag/src/types/status.rs:
```diff
@@ -0,0 +1,78 @@
+/// Defines the status of the Text Fragment search and extraction/search operation status
```
```suggestion
//! Defines the status of the Text Fragment search and extraction/search operation status
```

thiru-appitap replied:
refactored code takes are of this now!

mre commented on textfrag/src/types/url.rs:
```diff
@@ -0,0 +1,79 @@
+
+use crate::types::{FragmentDirective, FRAGMENT_DIRECTIVE_DELIMITER};
+
+/// Fragment Directive feature trait
```
```suggestion
/// Fragment Directive extension trait
```

thiru-appitap replied:
updated the comment

mre commented on textfrag/src/utils/mod.rs:
Maybe we don't need that file and can move the code closer to where it's used? (Assuming it's only used in one place.)

thiru-appitap replied:
removed the folder as part of the refactoring exercise

mre commented on textfrag/src/extract/fragdirtok.rs:
This module is so large, I'd assume a relatively large doc-block as well.
In there, I'd answer a few questions:
- What does the module do?
- How is it supposed to be used (including an example)?
- What were the design tradeoffs?
- What are the possible error conditions?

thiru-appitap replied:
working on it - will commit this documentation for review!

mre commented on textfrag/src/extract/fragdirtok.rs:
```diff
@@ -0,0 +1,644 @@
+
+use crate::types::{TextDirective, TextDirectiveKind, TextDirectiveStatus};
+
+const BLOCK_ELEMENTS: &[&str] = &[
```
I'm still not too happy that we have to maintain this list. 
Looked around for alternatives, and I found https://github.com/servo/html5ever/blob/main/html5ever/src/tree_builder/tag_sets.rs.
Not sure if it can be used, but I wanted to mention it.
If it's too much hassle to integrate, we can keep the current implementation.

thiru-appitap replied:
unfortunately, the html5ever does not expose these macros for us to consume and so couldn't use it - i don't want to keep a copy of this file inside lychee repository and so ruling it out. And i did search around to check on alternative approaches and found either we need to have a headless browser to manage/identify the element type (as block) or stay with this current approach. 
My recommendation, for now, is to keep this code - in future, if the html5ever makes these macros as public, we can potentially be freed up of maintaining this list.

mre commented:
I've added a few more comments.
Most of them are minor. I think it should be fairly easy to go through them and accept/reject my suggestions from the GitHub UI.

As for the error, it's currently failing because there is no `textfrag` crate on crates.io. That is to be expected, because we haven't published it yet. However, the question is if we *do* want to publish it at all. 
`cargo` expects all dependencies to be published on crates.io. We have three options:

1. Publish the crate under `lychee-textfrag`. This would require renaming the crate to fit our naming scheme.
2. Publish the crate as a separate project, e.g. `textfrag`. Then we'd have to move the code out into a separate repo and perhaps add a CI/CD process for releases. We could keep it in the `lycheeverse` namespace, or you publish it under your own name. The license has to be compatible (Apache/MIT).
3. Don't publish the crate. In that case, we'd have to move the `textfrag` crate as a submodule into the `lychee-lib` crate. The advantage is that we don't have to set up a separate repo, which will make future changes easier.

I'd vote for option 3, which is the easiest right now and allows us to keep maintaining the code inside lychee. It's still reasonably well encapsuled in its own module and we can always make it a separate crate later once the code is mature enough. The downside is longer compile-times because it would be in the same codegen unit as the rest of the library code.

<details>
<summary>Detailed instructions for option 3</summary>

To merge the `textfrag` code directly into `lychee-lib`, here's what you could do:

1. Move the `textfrag` code into `lychee-lib`, perhaps in a submodule like `lychee-lib/src/textfrag/`
2. Remove `textfrag` from your workspace members in the root `Cargo.toml`
3. Update any imports in `lychee-lib` to reference the new module location instead of the external crate

This approach has several benefits:
- Simplifies your publishing process - no need to maintain and publish a separate crate
- Makes it clear that this is internal implementation code
- Gives you more flexibility to change the code without worrying about breaking other potential users
</details>

Let me know what you think.

thiru-appitap commented:
> I've added a few more comments. Most of them are minor. I think it should be fairly easy to go through them and accept/reject my suggestions from the GitHub UI.
> 
> As for the error, it's currently failing because there is no `textfrag` crate on crates.io. That is to be expected, because we haven't published it yet. However, the question is if we _do_ want to publish it at all. `cargo` expects all dependencies to be published on crates.io. We have three options:
> 
> 1. Publish the crate under `lychee-textfrag`. This would require renaming the crate to fit our naming scheme.
> 2. Publish the crate as a separate project, e.g. `textfrag`. Then we'd have to move the code out into a separate repo and perhaps add a CI/CD process for releases. We could keep it in the `lycheeverse` namespace, or you publish it under your own name. The license has to be compatible (Apache/MIT).
> 3. Don't publish the crate. In that case, we'd have to move the `textfrag` crate as a submodule into the `lychee-lib` crate. The advantage is that we don't have to set up a separate repo, which will make future changes easier.
> 
> I'd vote for option 3, which is the easiest right now and allows us to keep maintaining the code inside lychee. It's still reasonably well encapsuled in its own module and we can always make it a separate crate later once the code is mature enough. The downside is longer compile-times because it would be in the same codegen unit as the rest of the library code.
> 
> Detailed instructions for option 3
> To merge the `textfrag` code directly into `lychee-lib`, here's what you could do:
> 
> 1. Move the `textfrag` code into `lychee-lib`, perhaps in a submodule like `lychee-lib/src/textfrag/`
> 2. Remove `textfrag` from your workspace members in the root `Cargo.toml`
> 3. Update any imports in `lychee-lib` to reference the new module location instead of the external crate
> 
> This approach has several benefits:
> 
> * Simplifies your publishing process - no need to maintain and publish a separate crate
> * Makes it clear that this is internal implementation code
> * Gives you more flexibility to change the code without worrying about breaking other potential users
> 
> Let me know what you think.

@mre I understand and agree with your recommendation - I am addressing this and, as well, the other review comments and republish for review - thanks for your patience!

mre commented:
@thiru-appitap, I saw that you did some work. Can you close the resolved conversations already? It becomes a bit hard to keep track of the open TODOs. 😉 
Also, did you use the "commit suggestion" feature from GitHub? For some of the changes, I made suggestions, which are easy to merge with a single click. This way, changes should be much easier to handle on your end. Of course, you have to `git pull` the changes locally to be up-to-date.

thiru-appitap commented on textfrag/src/types/directive/fragdirective.rs:
```diff
@@ -0,0 +1,336 @@
+    }
+
+    /// Extract Fragment Directive from the (fragment) string as input
+    /// Returns a list of the Text Directives from the fragment string
```
modified the comments

thiru-appitap commented:
> @thiru-appitap, I saw that you did some work. Can you close the resolved conversations already? It becomes a bit hard to keep track of the open TODOs. 😉 Also, did you use the "commit suggestion" feature from GitHub? For some of the changes, I made suggestions, which are easy to merge with a single click. This way, changes should be much easier to handle on your end. Of course, you have to `git pull` the changes locally to be up-to-date.

@mre was juggling between couple of priorities, along with local travels and so couldn't resolve & commit the changes earlier itself - now on I should be able to get back with much faster turnaround on the feedback
now, the changes are all done (fingerscrossed) based on the feedback; along with, I have tried to adopt the idiomatic way as much as I could - willing to learn more and please keep helping me with more feedback!

almereyda commented:
This branch kindly asks for a rebase.

Running `cargo install --branch text-fragment --git https://github.com/thiru-appitap/lychee.git` and then using `lychee --include-text-fragments` shows that this works in principle.

Would be nice to see this arrive.

thiru-appitap commented:
> This branch kindly asks for a rebase.
> 
> Running `cargo install --branch text-fragment --git https://github.com/thiru-appitap/lychee.git` and then using `lychee --include-text-fragments` shows that this works in principle.
> 
> Would be nice to see this arrive.

I have rebased the branch to the latest but am facing lint failures - working on to fix it and recommit by this weekend.

mre commented:
I guess you can just 

```sh
git fetch origin
git rebase origin/master
```

to get rid of the lint errors.

thomas-zahner commented:
Thank you @thiru-appitap for your great efforts. I started to overhaul this PR in https://github.com/thomas-zahner/lychee/tree/text-fragment-v2. However, in the meantime #2138 was created which is actually more mature already. Because of this I'd like to focus on #2138 which will probably soon be ready and merged. Feel free to take a look at the new PR. @Marzal Feel free to test the new PR.
